File: Easy/JAVASCRIPT/BranchSumBT.js
class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

function branchSum(node) {
    const sum = []
    branchSumHelper(node, 0, sum);
    return sum;
}

function branchSumHelper(node, currentSum, sum) {
    if (node == null) return;

    let endSum = currentSum + node.value;
    if (node.left == null && node.right == null) {
        sum.push(endSum);
        return;
    }

    branchSumHelper(node.left, endSum, sum)
    branchSumHelper(node.right, endSum, sum)
}


const tree = new Node(10);
tree.left = new Node(5);
tree.left.left = new Node(2);
tree.left.left.left = new Node(1);
tree.left.right = new Node(5);
tree.right = new Node(15);
tree.right.left = new Node(13);
tree.right.left.right = new Node(14);
tree.right.right = new Node(22);


const result = branchSum(tree);
console.log(result);

File: Easy/PYTHON/BranchSumBT.py
# Initialize the class input of BST
class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def get_branch_sum(tree):
    sum = []
    get_branch_sum_helper(tree, 0, sum)
    return sum


def get_branch_sum_helper(tree, runningSum, sum):
    if tree is None:
        return

    endSum = runningSum + tree.value
    if tree.left is None and tree.right is None:
        sum.append(endSum)
        return

    get_branch_sum_helper(tree.left, endSum, sum)
    get_branch_sum_helper(tree.right, endSum, sum)


def node_depths(root):
    sumOfDepths = 0
    stack = [{"node": root,  "height": 0}]
    while len(stack) > 0:
        stackItem = stack.pop()
        (node, height) = stackItem["node"], stackItem["height"]
        if node is None:
            continue
        sumOfDepths += height
        stack.append({"node": node.left,  "height": height + 1})
        stack.append({"node": node.right, "height": height + 1})
    return sumOfDepths


root = BST(10)
root.left = BST(5)
root.left.left = BST(2)
root.left.left.left = BST(1)
root.left.right = BST(5)
root.right = BST(15)
root.right.left = BST(13)
root.right.left.right = BST(14)
root.right.right = BST(22)

# result = get_branch_sum(root)
# print(result)

result = node_depths(root)
print(result)
